corehq/apps/ota/views.py

```diff
@@ -169,7 +169,14 @@ def phone_holds_all_cases(request):
         except MissingSyncLog:
             return False
 
-    if not case_ids_to_claim and phone_holds_all_cases(request):
+    def cases_to_claim_modified_since_last_synclog_date():
+        try:
+            synclog = get_properly_wrapped_sync_log(request.last_sync_token)
```

Comment:
Instead of re-fetching the synclog (which is quite an expensive operation) it would be better to extract this out of both the functions. An easy way to do this would be to extract it into an inner function and decorate it with `@memoized`. This will allow you to call the function multiple times without having to re-fetch the synclog, but since it's an inner function the memoized data won't persist beyond the outer function call:

```python
def claim(request, domain):

    @memoized
    def get_synclog():
        ...

    if get_synclog():
        synclog = get_synclog()

```

Reply:
[3ad5908](https://github.com/dimagi/commcare-hq/pull/32554/commits/3ad59083b81b42f8677fa2afa2055da12accc8c1)